ion-schema/src/ordered_elements_nfa.rs

```diff
@@ -0,0 +1,443 @@
+    /// outgoing edges must have consecutively numbered destinations—so we can model the edges as a
+    /// map of `from_id` (the index in the vec) to a `Range` of `StateId`.
+    edges: Vec<RangeInclusive<usize>>,
+    /// Stores the [StateId] of the final state (packaged into a [StateVisitCount]) for convenience.
+    terminal_state: StateVisitCount,
+}
+
+impl OrderedElementsNfa {
+    /// Constructs an [OrderedElementsNfa] from a [Vec] of pairs of [VariablyOccurringTypeRef] and
+    /// an optional string description.
+    ///
+    /// If no description is provided, a default is used. If a description is provided, it should
+    /// be something that is recognizable to users, such as a row/col in the source ISL or a snippet
+    /// of the source ISL.
+    pub fn new(intermediate_states: Vec<(VariablyOccurringTypeRef, Option<String>)>) -> Self {
+        // "Initial" state is always first—no surprise there.
+        let mut states = vec![State::Initial];
+
+        // Construct intermediate states and add to our vec of states.
+        intermediate_states
+            .into_iter()
+            .enumerate()
+            .for_each(|(i, (var_type_ref, description))| {
+                let description = description.unwrap_or(format!("<ORDERED_ELEMENT[{}]>", i));
+                let (min_visits, max_visits) = var_type_ref.occurs_range().inclusive_endpoints();
+                let state = IntermediateState {
+                    type_ref: var_type_ref.type_ref(),
+                    min_visits,
+                    max_visits,
+                    description,
+                };
+
+                states.push(State::Intermediate(i + 1, state))
+            });
+
+        // This will become the ID of the "Final" state, but it's convenient to wait to add the
+        // "Final" state until we've determined the edges of the graph.
+        let max_id = states.len();
+
+        // Construct an adjacency list to represent the edges of the graph.
+        let mut edges = vec![];
+        for (i, s) in states.iter().enumerate() {
+            // Loop back to self, if max is > 1
+            let min_transition = if s.can_reenter(1) { i } else { i + 1 };
+
+            // Add transitions forward up to (including) the first type with a min occurs that is greater than 0
+            let mut j = i + 1;
+            while j < max_id {
+                let state_j = states.get(j).unwrap();
+                if !state_j.can_exit(0) {
+                    break;
+                }
+                j += 1;
+            }
+            edges.push(min_transition..=j)
+        }
+
+        // We make `Final` have fake edges to `Initial`. `Initial` denies all entry and `Final`
+        // doesn't allow exiting, so it's not actually traversable, and besides, you can't enter
+        // `Final` unless you're at the end-of-sequence event, so there's no more input to handle
+        // that would make it even try any further transitions.
```

Comment:
It turns out I don't actually need them. I added them because I'm used to thinking about the way arrays work in Java, and in that case it simplifies things to have a fake entry so that the `edges` and `states` have the same length. However, a `Vec::get` returns an `Option`, so I don't actually need them here.